Require scheduling infra components onto control-plane nodes

iholder101 opened the pull request:
### What this PR does
Before this PR:
Some of Kubevirt's infrastructure components are powerful and privileged. Running these components on worker nodes is a security issue which compromises the cluster. For more information, please see the conversation here: https://github.com/kubevirt/community/pull/246#discussion_r1550188378

As of today, it's solely the admin's responsibility to set the right NodePlacement under Kubevirt CR. It seems that some major Kubevirt users (e.g. OpenShift Virtualization) are not setting this properly, and we should strive to help them by using a sensible default.

After this PR:
Kubevirt, by default, requires running infrastructure components on control-plane nodes. This will only happen if the user did not specify any node placement settings in Kubevirt CR.

#### Special notes for your reviewer:
After this PR Kubevirt infra components won't be able to schedule to worker nodes, unless specified otherwise on Kubevirt CR. This might cause some environments to fail if they won't adapt their Kubevirt CR properly.

**Archived:**
~The `InjectPlacementMetadata` function is being applied to all of our infra components, including virt-handler, which (as a daemonset) is expected to be scheduled to every node. Bear in mind that defining a `PreferredDuringSchedulingIgnoredDuringExecution` for a daemonset does **not** mean it wouldn't be scheduled onto every node, but instead this preference would just be ignored.~

~Therefore, we can still apply this function onto every infra components without treating virt-handler as a special case.~

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] ~Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required~
- [X] PR: The PR description is expressive enough and will help future contributors
- [X] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] ~Upgrade: Impact of this change on upgrade flows was considered and addressed if required~
- [X] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] ~Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.~
- [ ] ~Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered~

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Require scheduling infra components onto control-plane nodes
```

iholder101 commented:
FYI @rmohr

fossedihelm commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -520,5 +521,36 @@ var _ = Describe("Apply", func() {
+			}
+
+			Expect(masterPreferenceFound).To(BeTrue(), "preferring node-role.kubernetes.io/master is not injected to metadata")
+			Expect(controlPlanePreferenceFound).To(BeTrue(), "preferring node-role.kubernetes.io/master is not injected to metadata")
```
```suggestion
Expect(controlPlanePreferenceFound).To(BeTrue(), "preferring node-role.kubernetes.io/control-plane is not injected to metadata")
```

iholder101 replied:
Nice catch! Thanks!

fossedihelm reviewed (approved):
Thank you @iholder101 A note we should care about is that the `InjectPlacementMetadata` function is also used by the `virt-handler` daemonset. As long as we use `PreferredDuringSchedulingIgnoredDuringExecution` there are no problems, but it will be dangerous if, for any reason, in the future we want to switch to `RequiredDuringSchedulingIgnoredDuringExecution`, since this will implies that  virt-handler will not be scheduled on all the nodes but only on control-plane nodes.
Also, I think that we should prefer, playing with some weight, only `control-plane` nodes to both `control-plane,worker` nodes. WDYT?

fossedihelm commented:
/lgtm cancel
Wrong button :)

iholder101 commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -129,11 +129,43 @@ func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev
+								Preference: corev1.NodeSelectorTerm{
+									MatchExpressions: []corev1.NodeSelectorRequirement{
+										{
+											Key:      "node-role.kubernetes.io/master",
```
Cool!
I wasn't sure how long ago it was deprecated. Thanks for this data point!

iholder101 commented:
Thanks a lot for your review @fossedihelm!
I've addressed your suggested code changes :) PTAL

> A note we should care about is that the InjectPlacementMetadata function is also used by the virt-handler daemonset. 

You're absolutely right. I wanted to mention it in the PR description but forgot, I've added it now. Nice catch.

> As long as we use PreferredDuringSchedulingIgnoredDuringExecution there are no problems, but it will be dangerous if, for any reason, in the future we want to switch to RequiredDuringSchedulingIgnoredDuringExecution, since this will implies that virt-handler will not be scheduled on all the nodes but only on control-plane nodes.

I absolutely agree.
We should probably never, by default, define a node placement that requires scheduling on control-plane nodes. This is only a preference and it's only being added if the user did not specify any other node placement configs (including both requirements or preferences).

So to conclude, IMO this PR sets a more sensible default and brings no danger that I can see.

iholder101 commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -520,5 +521,39 @@ var _ = Describe("Apply", func() {
+
+			Expect(podSpec.Affinity).ToNot(BeNil())
+			Expect(podSpec.Affinity.NodeAffinity).ToNot(BeNil())
+			Expect(podSpec.Affinity.NodeAffinity.PreferredDuringSchedulingIgnoredDuringExecution).ToNot(BeNil())
```
Thanks!
The `HaveLen` check is being performed a few lines below

iholder101 replied:
Done!

iholder101 commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -520,5 +521,39 @@ var _ = Describe("Apply", func() {
 		})
+
+		It("should prefer scheduling to control-plane nodes", func() {
+			InjectPlacementMetadata(nil, podSpec)
```
Do you mean `not nil`?

iholder101 replied:
Added a test for the non-nil case :+1:

fossedihelm replied:
Great the addition of the non-nil nodePlacement.
Just a nit: here you are providing a nil `componentConfig`, while I was requesting a testcase for nil `componentConfig.NodePlacement` :)

Barakmor1 commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -129,11 +129,42 @@ func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev
+								},
+								Weight: 100,
+							},
+							{
+								Preference: corev1.NodeSelectorTerm{
+									MatchExpressions: []corev1.NodeSelectorRequirement{
+										{
+											Key:      "node-role.kubernetes.io/worker",
+											Operator: corev1.NodeSelectorOpDoesNotExist,
+										},
+									},
+								},
+								Weight: 99,
+							},
```
Why do we need affinity for the workers?

iholder101 replied:
@fossedihelm raised the concern that in some environments you can have nodes that are both workers and CPs. In this case, we'd prefer running on nodes that are **just** CP, and not worker+CP. It's a small nuance that might be relevant for some environments.

The reason for the weight being slightly lower is that it's a higher priority to run on CP nodes than avoid running on worker nodes.

Barakmor1 reviewed (commented):
Hey, Thanks for handling this :)

iholder101 commented:
Thanks @fossedihelm :) PTAL

fossedihelm reviewed (approved):
Thank you @iholder101!

iholder101 commented:
/test pull-kubevirt-unit-test 

passes locally. looks like flakiness.

fossedihelm commented:
> Thank you @iholder101 A note we should care about is that the `InjectPlacementMetadata` function is also used by the `virt-handler` daemonset. As long as we use `PreferredDuringSchedulingIgnoredDuringExecution` there are no problems, but it will be dangerous if, for any reason, in the future we want to switch to `RequiredDuringSchedulingIgnoredDuringExecution`, since this will implies that virt-handler will not be scheduled on all the nodes but only on control-plane nodes.

@iholder101 Seems like you switched to `RequiredDuringSchedulingIgnoredDuringExecution` and thus we fall into the situation described above :)

iholder101 commented:
> > Thank you @iholder101 A note we should care about is that the `InjectPlacementMetadata` function is also used by the `virt-handler` daemonset. As long as we use `PreferredDuringSchedulingIgnoredDuringExecution` there are no problems, but it will be dangerous if, for any reason, in the future we want to switch to `RequiredDuringSchedulingIgnoredDuringExecution`, since this will implies that virt-handler will not be scheduled on all the nodes but only on control-plane nodes.
> 
> @iholder101 Seems like you switched to `RequiredDuringSchedulingIgnoredDuringExecution` and thus we fall into the situation described above :)

Man, you're quick! :)
You're right, I was just thinking the same thing (and wonder how tests haven't failed yet?).

Unfortunately I think we must differentiate between daemon sets and deployments here.
Let me rework the PR.

fossedihelm commented:
> Man, you're quick! :) You're right, I was just thinking the same thing (and wonder how tests haven't failed yet?).
> 
> Unfortunately I think we must differentiate between daemon sets and deployments here. Let me rework the PR.

xD I saw `Require` and got triggered :D
Take your time :+1:

iholder101 commented:
FYI @jean-edouard @xpivarc @acardace @fossedihelm @Barakmor1 @rmohr @vladikr @fabiand @stu-gott 

The PR is now ready to review. I've re-worked it so that now we'd **require** landing our infra components on CP nodes, not just preferencing it.

Please refer to the PR description for more info. I'd like to emphasize this section however:

> After this PR Kubevirt infra components won't be able to schedule to worker nodes, unless specified otherwise on Kubevirt CR. This might cause some environments to fail if they won't adapt their Kubevirt CR properly.

This change might be a breaking change for some environments.
Please let me know if you have any concerns regarding this.

@dhiller do we run a single node cluster on CI? If it's labelled as both a worker and a CP node infra components wouldn't be schedulable. This is applicable to single-node deployments in general.

iholder101 commented:
I've tested offline, and it seems that kubevirt CI with `export KUBEVIRT_NUM_NODES=1` cannot be deployed with these changes.

I wonder if we want to postpone this PR until we fix CI issues, or we should:
1. Add this as a preference
2. Fix CI issues
3. Follow-up to change preferences into requirements

WDYT?

jean-edouard commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -125,15 +125,42 @@ const (
+												Key:      "node-role.kubernetes.io/control-plane",
+												Operator: corev1.NodeSelectorOpExists,
+											},
+											{
+												Key:      "node-role.kubernetes.io/worker",
+												Operator: corev1.NodeSelectorOpDoesNotExist,
+											},
```
This is wrong, control planes can be workers too, they're still control planes and can still schedule control-plane-related workloads.

iholder101 replied:
Right, but then how do we mitigate the security risk? IIUC the whole idea is to not schedule pods into worker nodes since it's dangerous.

What we can do is require scheduling to CP nodes, but prefer not landing on worker nodes.
Any objections?

iholder101 replied:
The reason I'm saying that is because there may be nodes on the clusters that are CP only, and some that are CP + worker. I think a sensible default is to prefer not landing on CP+worker nodes.

fossedihelm replied:
>What we can do is require scheduling to CP nodes, but prefer not landing on worker nodes.
Any objections?

:+1:

jean-edouard replied:
Preferring non-workers is a good idea.
The most important by far, in my opinion, is to avoid scheduling CP workloads on non-CP nodes.

iholder101 replied:
Done :)
PTAL

dhiller commented:
> @dhiller do we run a single node cluster on CI? If it's labelled as both a worker and a CP node infra components wouldn't be schedulable. This is applicable to single-node deployments in general.

The [periodic arm e2e runs with only one node](https://github.com/kubevirt/project-infra/blob/db0456785ebbabd944ec18c529a5b8ce79a810ce/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-periodics.yaml#L735), a quick search didn't yield any other jobs doing this.

acardace commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -552,14 +552,21 @@ var _ = Describe("Apply", func() {
 				Entry("nil node placement", &v1.ComponentConfig{}),
 			)
+
+			It("should not add requirement scheduling to control-plane nodes if the user already provided a node placement", func() {
```
I think this case can be merged with the `DescribeTable` above.

iholder101 replied:
TBH I'm not sure it'll make it easier to read or understand. The describe table deals with the case of adding the requirements (i.e. a positive test). Adding negative tests on the same table seems uncohesive to me.

Instead, I've added a dedicated context for negative tests. The tests there are completely straight-forward almost one-liner tests, and I'm not sure merging them into a table would do more good than damage

iholder101 replied:
Yeah, I guess it's a matter of preference :)
I just tend to think that super obvious and short test cases are easier to read with a dedicated test. But I see your point

iholder101 commented:
> > @dhiller do we run a single node cluster on CI? If it's labelled as both a worker and a CP node infra components wouldn't be schedulable. This is applicable to single-node deployments in general.
> 
> The [periodic arm e2e runs with only one node](https://github.com/kubevirt/project-infra/blob/db0456785ebbabd944ec18c529a5b8ce79a810ce/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-periodics.yaml#L735), a quick search didn't yield any other jobs doing this.

Thanks Daniel!
We now shifted to only preferring running on non-workers, so it should be fine. Thanks a lot!

iholder101 commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -125,7 +125,7 @@ const (
 
 // Merge all Tolerations, Affinity and NodeSelectos from NodePlacement into pod spec
-func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec) {
+func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec, defaultScheduleToCPNodes bool) {
```
Great suggestion! Done, thank!

acardace replied:
looks much better!

iholder101 commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -129,11 +129,47 @@ func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev
+				},
+			}
+		} else {
+			componentConfig = &v1.ComponentConfig{NodePlacement: &v1.NodePlacement{}}
```
Yeah, I thought about how to beautify it too, but I've tried playing with the code, and TBH it doesn't look better to me this way.

There's also the small overhead of creating new structs just to override them right afterwards (which should be pretty negligible, but still).

iholder101 commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -520,5 +521,45 @@ var _ = Describe("Apply", func() {
+					workerLabel       = "node-role.kubernetes.io/worker"
+				)
+
+				InjectPlacementMetadata(config, podSpec, true)
```
There's already a test for that

iholder101 commented on pkg/virt-operator/resource/apply/reconcile_test.go:
```diff
@@ -567,6 +567,11 @@ var _ = Describe("Apply", func() {
 				Expect(podSpec.Affinity).To(BeNil())
 			})
+
+			It("should not add requirement scheduling to control-plane nodes if the corresponding parameter is set to false", func() {
```
See comment [here](https://github.com/kubevirt/kubevirt/pull/11659#discussion_r1559370909)

acardace commented on pkg/virt-operator/strategy_job.go:
```diff
@@ -98,7 +98,7 @@ func (c *KubeVirtController) generateInstallStrategyJob(infraPlacement *v1.Compo
 	}
 
-	apply.InjectPlacementMetadata(infraPlacement, &job.Spec.Template.Spec)
+	apply.InjectPlacementMetadata(infraPlacement, &job.Spec.Template.Spec, apply.AnyNode)
```
isn't this a mistake? Shouldn't you pass `RequireControlPlanePreferNonWorker` here?

iholder101 replied:
Yeah, I think you're right. Good catch! done

acardace commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -125,15 +132,51 @@ const (
 	if componentConfig == nil || componentConfig.NodePlacement == nil {
-		componentConfig = &v1.ComponentConfig{
-			NodePlacement: &v1.NodePlacement{},
+		if nodePlacementOption == RequireControlPlanePreferNonWorker {
```
nit: I always like using switch cases more with enums rather than if-else statement, just a personal preference

iholder101 replied:
Yeah I agree :) will change that, thank you!

iholder101 replied:
Done.
Please note that I've added a default case that panics if an unknown `nodePlacementOption` is provided. I think it's better to fail fast in this case, and since this value is internal and not user-facing I think it's the right choice here. Any thoughts?

acardace replied:
I don't think panicking is ever a good idea in production code, can't you just default to a sane behavior in the `default` case?

acardace replied:
Logging and unit tests are good tools to check that passing unknown values cannot happen, I'd strongly avoid the panic.

iholder101 replied:
> I don't think panicking is ever a good idea in production code

I think that's certainly right in most cases. However my concern is that a bug would cause an unexpected behavior that would be hard to spot out. After all this function is completely internal, wouldn't we want to spot the bug right away if an unknown value is being provided? If it was a user-facing function I would speak differently.

> Logging and unit tests are good tools to check that passing unknown values cannot happen

Unit tests? Can you explain how they're helpful here?

w.r.t. logging - Despite easing the debug process a bit, I think the probability of merging a bad PR, causing a surprising behavior, and not being able to spot it for a while is still high.

acardace replied:
> I think that's certainly right in most cases. However my concern is that a bug would cause an unexpected behavior that would be hard to spot out. After all this function is completely internal, wouldn't we want to spot the bug right away if an unknown value is being provided? If it was a user-facing function I would speak differently.

I think that's one more reason to not panic, panics should be exercised when the program execution cannot continue (fatal error), one good example is segmentation faults, where the program panics because otherwise, it would access non-mapped memory. I don't think this is a good case for panicking, you can just pick a sane default and continue the execution without disruptions.

> Unit tests? Can you explain how they're helpful here?

Fuzzy testing, just pass some random values into the function and test that if the supplied input is invalid it falls back to a default case (which is placement on control plane).

> w.r.t. logging - Despite easing the debug process a bit, I think the probability of merging a bad PR, causing a surprising behavior, and not being able to spot it for a while is still high.

There's no surprising runtime behavior if you fallback to a sane default, actually, you just sanitized input.

It's way more disrupting to panic in this case than just opt for a safe and sane default, I honestly don't see the case for panicking here.

iholder101 replied:
I've removed the panic. PTAL.

acardace commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -57,6 +57,13 @@ import (
 const Duration7d = time.Hour * 24 * 7
 const Duration1d = time.Hour * 24
 
+type DefaultInfraComponentsNodePlacementOption int
```
Can you change this to `InfraComponentsNodePlacementOption`? It's a bit unusual to have the word `Default` in a type name.

iholder101 replied:
Yeah this was the original name.
However I wanted to emphasize that we're setting a default option here, as if the node-placement is already set (even without node affinity) we won't mutate it. I think that setting a `InfraComponentsNodePlacementOption` to require CP nodes, then get pods that are not requiring CP nodes, is confusing. But setting a `DefaultInfraComponentsNodePlacementOption` might be clearer.

WDYT?

acardace replied:
I'd prefer `InfraComponentsNodePlacementOption`, Default in a type name is a bit confusing as a value can be the default or not for a type but the type itself cannot be default or not(?).

The fact that one of the options is supplied by default is anyway pretty clear from the code, or you can make it even clearer with a comment just before defining the default value.

acardace replied:
You can also alias the default value to make it clearer.

iholder101 replied:
> I'd prefer InfraComponentsNodePlacementOption, Default in a type name is a bit confusing as a value can be the default or not for a type but the type itself cannot be default or not(?).

A default name in a type is not confusing IMO if the type itself represents a default value. Quite the opposite - it reflects the purpose of the type. A value that is of type `Default..` is a default value for something else, obviously, not for the type itself. I just kind of miss why it is confusing.

BTW, types are being prefixed with default pretty often in k8s, but also in Kubevirt:
```bash
> grep -E -i "^type default" ./kubernetes -R --include=*.go --exclude-dir=vendor | wc -l
54

> grep -E -i "^type default" ./kubevirt -R --include=*.go --exclude-dir=vendor | wc -l
5
```

> The fact that one of the options is supplied by default is anyway pretty clear from the code, or you can make it even clearer with a comment just before defining the default value.

I strongly disagree with this approach.
In order to reason about a function and its parameters you only need to look at the function signature. Having to look at the code is a huge code smell, and having to document such a code with comments is even worse IMHO.

fossedihelm replied:
I tend to agree with @acardace honestly.
If we want to emphasize that it acts as default, it should be the function parameter name that has the `default` (not the type).
Also, what will happen if we decide in the future that this will not be a default? Should we change the type name? Just my cents :)

fossedihelm replied:
IOW, how about use `InfraComponentsNodePlacementOption` and change the parameter name of `InjectPlacementMetadata` to `defaultNodePlacementOption`?

xpivarc replied:
> IOW, how about using `InfraComponentsNodePlacementOption` and changing the parameter name of `InjectPlacementMetadata` to `defaultNodePlacementOption`?

+1

iholder101 replied:
> Both function signatures with or without the default in the type clearly say how the parameter will be used.

Can you explain? How would the signature reflect how the parameter will be used without the default?

> If we want to emphasize that it acts as default, it should be the function parameter name that has the default (not the type).

Why it shouldn't be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.

I'm sorry but I'm really not convinced :) and find it hard to understand your reasoning for why it is confusing. I mean, you claim that it's confusing but don't really explain why. The reasoning for why it is valuable is very simple - it reflects that the type's values serve as default values only. I still can't find a good reason for not reflecting this information, or reflecting it only at the parameter name level.

> Anyway, I think we agree to disagree :)

Sure :)

iholder101 replied:
FYI, changed `DefaultInfraComponentsNodePlacementOption` to `DefaultInfraComponentsNodePlacement` (without the option in the end) to make it shorter and more readable.

iholder101 replied:
Sorry @fossedihelm, I just now realized I haven't responded to one of your arguments.

> Also, what will happen if we decide in the future that this will not be a default?

What do you mean we'll decide it won't be the default? In that case the type can be deleted, as it's no longer necessary, since its whole purpose is to define a default node placement. Can you elaborate on this one? Which other use we can make of this type in your opinion?

> Should we change the type name?

I don't see why we would ever change it as explained above.

But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.

fossedihelm replied:
>Why shouldn't it be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.

I think that is here where we disagree. The purpose of the type is to represent the kind of a particular struct.
It is in the `InjectPlacementMetadata` that it takes the form of default value. Outside of the function scope, it is not a default value. You said that it represents a default value, but, thinking outside of the `InjectPlacementMetadata`, the default value of what?
IMO, it becomes a default value when used inside the `InjectPlacementMetadata`, and only there; for this reason to me it is better to change the parameter name.
The same will happen for every structs. We have vms, pods, etc... These types can be used as default for something in some circumstances. But we will change the parameter name, not the type because a pod will always be a pod.
This is my reasoning :) 

>But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.

Again, a type should describe the type(kind or whatever), not the behavior :)

Maybe we just have different opinions about it.

acardace replied:
> Can you explain? How would the signature reflect how the parameter will be used without the default?

`DefaultInfraComponentsNodePlacementOption` vs `InfraComponentsNodePlacementOption`, I just understand a param of this type will be used for the node placement of infra components, `Default` in the type doesn't convey anything, what should default mean? I'm passing an explicit value, how does `Default` tell me how the function will behave? My point is, it's not clear one way or the other.

We're full of examples in our APIs where default values are documented with comments and this is because there's no such thing as a default type in a statically compiled language there can only be default values, but those depend on how the function is used and the value semantics. A default type by itself is a contradiction.

I think what you want to mean is that this value will be used by default if no explicit `componentConfig` is passed in, but the `Default` in the type doesn't tell me that, when I read a type name it must tell me something about the type itself not about how some function using that type should behave when another param is nil.

I think the function is ill-defined, why pass both `componentConfig` and `nodePlacementOption` explicitly? They refer to the same thing, there's a cross dependency that is not clear whether you put the `Default` in the type name or not.  One of the issue is that this function is both used for deployments and daemonsets, and this is why you need the additional parameter, otherwise you wouldn't, you could just encode in the function that if `componentConfig` then the default behavior is `X` or `Y`.

Anyway as I said, even if I don't agree I'm not going to block this.

iholder101 replied:
> > Why shouldn't it be part of the type name if the whole purpose of the type is to represent a default value? If we'd only change the parameter name we make this visible only to the function signature, not the type as a whole.
> 
> I think that is here where we disagree. The purpose of the type is to represent the kind of a particular struct. It is in the `InjectPlacementMetadata` that it takes the form of default value. Outside of the function scope, it is not a default value. You said that it represents a default value, but, thinking outside of the `InjectPlacementMetadata`, the default value of what? IMO, it becomes a default value when used inside the `InjectPlacementMetadata`, and only there; for this reason to me it is better to change the parameter name.

This enum is strictly for the purpose of defining a default node placement, hence its name. It does not, and should not, have any other meaning. Outside of the `InjectPlacementMetadata` function it could be used but only for this specific meaning.

You can't just refer to the values themselves without considering their type and then make use of them in a different context. This is why golang supports type branding (i.e. creating new type names out of existing types, like I did here for `int`) in the first place.

In Kubernetes, for example, the value `"Exists"` is defined as part of different types:
```go
TolerationOpExists TolerationOperator = "Exists"
```
```go
ScopeSelectorOpExists       ScopeSelectorOperator = "Exists"
```
```go
NodeSelectorOpExists       NodeSelectorOperator = "Exists"
```

I think that following your reasoning, they could have just used a value for any purpose. But the reason for doing so is to give a value a very specific context in which it is relevant. You can't use `TolerationOpExists` instead of `NodeSelectorOpExists` just because the underlying value is the same. In the same way, you can't just use `AnyNode` to represent anything you want - it is specifically dedicated for default node placement values.

[1] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L2934
[2] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L3223
[3] https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/pkg/apis/core/types.go#L5780

> The same will happen for every structs. We have vms, pods, etc... These types can be used as default for something in some circumstances. But we will change the parameter name, not the type because a pod will always be a pod. This is my reasoning :)

The purpose of vms and pods is not to serve as default values. Their usage is wide. This is not true for the type I'm introducing.

> > But if there's a hypothetical need to change the name of a completely internal type, are you trying to imply that it's a bad thing to do? Can you explain why or what your concerns are? IMHO doing so is completely valid and again I'm not sure exactly what's the concern / risk / problem and the reasoning for it.
> 
> Again, a type should describe the type(kind or whatever), not the behavior :)
> 
> Maybe we just have different opinions about it.

If that's true, can you give me a reason for doing the following?

```go
type MyCoolType int
```

The type is int. Why would we want to re-brand it with a different name if not to describe the usage of this type for a specific context?

iholder101 replied:
> > Can you explain? How would the signature reflect how the parameter will be used without the default?
> 
> `DefaultInfraComponentsNodePlacementOption` vs `InfraComponentsNodePlacementOption`, I just understand a param of this type will be used for the node placement of infra components, `Default` in the type doesn't convey anything, what should default mean? I'm passing an explicit value, how does `Default` tell me how the function will behave? My point is, it's not clear one way or the other.

The default emphasizes that these values are relevant specifically in the context of providing a default node placement. So it lets you understand something you wouldn't have understood otherwise, and even worse, you might end up using it for a different purpose which its not designed for.

> We're full of examples in our APIs where default values are documented with comments and this is because there's no such thing as a default type in a statically compiled language there can only be default values, but those depend on how the function is used and the value semantics. A default type by itself is a contradiction.

It's not a default type, it a type of default values. I'm not sure what you mean that there is "no such thing" in compiled languages. Go supports type branding specifically to re-brand the same type but give it a concrete context, and that's what I'm doing here. Go then lets you request a type of that specific re-branded type, so it's not operating solely on the basis of value semantics. Further to that, I previously showed that Kubernetes uses this pattern for about 60 times in their code base. I'm not sure why you're picturing it as contradictory or as something that's obviously wrong.

> I think what you want to mean is that this value will be used by default if no explicit `componentConfig` is passed in, but the `Default` in the type doesn't tell me that, when I read a type name it must tell me something about the type itself not about how some function using that type should behave when another param is nil.

It's not about how "some function is using it" - it is the sole purpose of this type and why it exists. The fact that it's being used by a certain function is an implementation detail, but nevertheless the type should inform you that what it represents is a set of default values for node placement and shouldn't be used for anything else, no matter in which function.

> I think the function is ill-defined, why pass both `componentConfig` and `nodePlacementOption` explicitly? They refer to the same thing, there's a cross dependency that is not clear whether you put the `Default` in the type name or not. One of the issue is that this function is both used for deployments and daemonsets, and this is why you need the additional parameter, otherwise you wouldn't, you could just encode in the function that if `componentConfig` is nill then the default behavior is `X` or `Y`.

Right. If it was for deployments I wouldn't have needed it.
But the reason for passing both `componentConfig` and `nodePlacementOption` is because this function is used for many different objects. For most objects you'd want a way of declaring how to default node placement (i.e., as default means, only if it's not already specified by the `componentConfig`)

> Anyway as I said, even if I don't agree I'm not going to block this.

:+1:

fossedihelm replied:
As I said, I think we just have different opinions :)
For sure, nothing that will block this PR from getting merged.

iholder101 replied:
Thanks @acardace and @fossedihelm for the discussion and for the elaborations :heart:

iholder101 commented on pkg/virt-operator/resource/apply/apps.go:
```diff
@@ -276,7 +276,7 @@ func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) (bool, error) {
 	injectOperatorMetadata(kv, &daemonSet.ObjectMeta, imageTag, imageRegistry, id, true)
 	injectOperatorMetadata(kv, &daemonSet.Spec.Template.ObjectMeta, imageTag, imageRegistry, id, false)
-	InjectPlacementMetadata(kv.Spec.Workloads, &daemonSet.Spec.Template.Spec)
+	InjectPlacementMetadata(kv.Spec.Workloads, &daemonSet.Spec.Template.Spec, AnyNode)
```
Not in this case.
DaemonSets need to be deployed on every node, not only control plane nodes.

acardace commented on pkg/virt-operator/resource/apply/apps.go:
```diff
@@ -51,7 +51,7 @@ func (r *Reconciler) syncDeployment(origDeployment *appsv1.Deployment) (*appsv1.
 	injectOperatorMetadata(kv, &deployment.ObjectMeta, imageTag, imageRegistry, id, true)
 	injectOperatorMetadata(kv, &deployment.Spec.Template.ObjectMeta, imageTag, imageRegistry, id, false)
-	InjectPlacementMetadata(kv.Spec.Infra, &deployment.Spec.Template.Spec)
+	InjectPlacementMetadata(kv.Spec.Infra, &deployment.Spec.Template.Spec, AnyNode)
```
Shouldn't this be `RequireControlPlanePreferNonWorker` ?

iholder101 replied:
Good catch!! Thanks!

acardace commented:
/approve

Even if there is a small detail I don't agree with in the implementation, it is anyway important that this PR gets merged.

Thanks @iholder101 for the PR and the patience.

fossedihelm reviewed (approved):
I was forgetting about this PR. Thank you!

iholder101 commented:
/test pull-kubevirt-e2e-kind-sriov

iholder101 commented:
So it seems that `pull-kubevirt-e2e-kind-sriov` fails because the only CP node is tainted:
```yaml
"taints": [
    {
        "key": "node-role.kubernetes.io/control-plane",
        "effect": "NoSchedule"
    }
]
```

@acardace @fossedihelm @jean-edouard I guess we should add a toleration.
Do you think we should add a "wildcard" toleration (i.e. for every `NoSchedule` taint)? Or only tolerrate this specific `control-plane` key taint?

iholder101 commented:
/hold

to avoid bot re-tests

xpivarc commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -125,15 +132,57 @@ const (
+								NodeSelectorTerms: []corev1.NodeSelectorTerm{
+									{
+										MatchExpressions: []corev1.NodeSelectorRequirement{
+											{
```
It seems this label is not applied to upgraded clusters. At least we need to document this.

iholder101 replied:
After looking into it, it seems that in some cases this label can be missing.
In fact, I've initially implemented the PR in a way that it requires either a CP or a Master label, but @fossedihelm [asked to](https://github.com/kubevirt/kubevirt/pull/11659#discussion_r1555270580) remove it.

@fossedihelm - IMHO we can be on the safe side and add both labels. TBH I don't think there's anything to lose from using such approach. Do you have any objections?

fossedihelm replied:
Kinda weird! But agree to cover also this! Sorry for the wrong suggestion

fossedihelm replied:
>- Introduce a new label "node-role.kubernetes.io/control-plane" that will be applied in parallel to "node-role.kubernetes.io/master" until the removal of the "node-role.kubernetes.io/master" label.
> - Make "kubeadm upgrade apply" add the "node-role.kubernetes.io/control-plane" label on existing nodes that only have the "node-role.kubernetes.io/master" label during upgrade.

From 1.20 k8s changelog https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md :thinking:

ormergi replied:
Asking to understand, why does Kubevirt should support the Master label if its deprecated on 1.24 while Kubevirt support k8s 1.29,8,7 at the moment?

iholder101 replied:
Why not? :)

As said above, in some environments the master label is still applicable (e.g. due to upgrade bugs). We have no real reason to avoid being on the safe side IMO.

iholder101 commented:
> So it seems that `pull-kubevirt-e2e-kind-sriov` fails because the only CP node is tainted:
> 
> ```yaml
> "taints": [
>     {
>         "key": "node-role.kubernetes.io/control-plane",
>         "effect": "NoSchedule"
>     }
> ]
> ```
> 
> @acardace @fossedihelm @jean-edouard I guess we should add a toleration. Do you think we should add a "wildcard" toleration (i.e. for every `NoSchedule` taint)? Or only tolerrate this specific `control-plane` key taint?

After an offline discussion we reached a consensus that we do not need to tolerate taints by default. The admin can always add tolerations to Kubevirt CR's node placement configurations. Instead we need to configure the failing lane to not taint the CP node, unless there's a reason not to. FYI @dhiller.

I've opened https://github.com/kubevirt/project-infra/pull/3352 in order to address that.

iholder101 commented:
https://github.com/kubevirt/kubevirtci/pull/1173 is now merged.
Waiting for https://github.com/kubevirt/kubevirt/pull/11786 (which brings this PR to Kubevirt) to get in and then I'll rebase upon it.

iholder101 commented:
/test pull-kubevirt-e2e-kind-1.27-vgpu
/test pull-kubevirt-e2e-kind-sriov

iholder101 commented:
/test pull-kubevirt-e2e-kind-1.27-vgpu

iholder101 commented:
kind lanes have passed now :tada: 
/unhold

@fossedihelm would you mind re-LGTMing? I've rebased, no further code changes.

fossedihelm reviewed (approved):
Thank you!

fossedihelm commented:
/hold
@iholder101, could you please check if the network failure is related to this PR?
Feel free to unhold otherwise

iholder101 commented:
Thanks @fossedihelm!
I don't see how the failures are related to this PR, but they are recurring.

@ormergi @EdDev would you mind taking a look?

ormergi commented:
hey @iholder101,  I went through few  of the failing sig-net jobs artifacts, very few test fail (1-2) due to VMs not being ready on time.
Looking deeper we can see repeating issue the accommodating node of the failing VMs, due to what seems to be issue with kubelet.

For example: [Job](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784392104701595648), test: 
 `[sig-network] [Serial] VirtualMachineInstance with passt network binding plugin should allow regular network connection Connectivity should be preserved after migration [IPv4]`  
Failed because the vmi  `testvmi-vbdnq` was not ready on time.

Looking into [event.log](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784392104701595648/artifacts/k8s-reporter/1_events.log), we can see an event saying the node that accommodate this VMI, "node2",  is not ready:
```
{
            "metadata": {
                "name": "virt-launcher-testvmi-vbdnq-jnhrf.17ca5380314fb2a9",
                "namespace": "kubevirt-test-default1",
                "uid": "384e7f19-e60d-4bbb-bd14-e289394240bc",
                "resourceVersion": "74045",
                "creationTimestamp": "2024-04-28T03:20:18Z",
                "managedFields": [
...
            },
            "involvedObject": {
                "kind": "Pod",
                "namespace": "kubevirt-test-default1",
                "name": "virt-launcher-testvmi-vbdnq-jnhrf",
                "uid": "08012b52-12f7-4480-a3ba-b63187171201",
                "apiVersion": "v1",
                "resourceVersion": "73830"
            },
            "reason": "NodeNotReady",
            "message": "Node is not ready",
            "source": {
                "component": "node-controller"
            },
            "firstTimestamp": "2024-04-28T03:20:18Z",
            "lastTimestamp": "2024-04-28T03:20:18Z",
            "count": 1,
            "type": "Warning",
            "eventTime": null,
            "reportingComponent": "node-controller",
            "reportingInstance": ""
        },
```

Looking into [node.log](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784392104701595648/artifacts/k8s-reporter/1_nodes.log) we can see "node02" is not ready due what seems to be related to kubelet not being ready:
```
                "conditions": [
                    {
                        "type": "NetworkUnavailable",
                        "status": "False",
                        "lastHeartbeatTime": "2024-04-28T01:24:32Z",
                        "lastTransitionTime": "2024-04-28T01:24:32Z",
                        "reason": "CalicoIsUp",
                        "message": "Calico is running on this node"
                    },
                    {
                        "type": "MemoryPressure",
                        "status": "Unknown",
                        "lastHeartbeatTime": "2024-04-28T03:19:05Z",
                        "lastTransitionTime": "2024-04-28T03:20:17Z",
                        "reason": "NodeStatusUnknown",
                        "message": "Kubelet stopped posting node status."
                    },
                    {
                        "type": "DiskPressure",
                        "status": "Unknown",
                        "lastHeartbeatTime": "2024-04-28T03:19:05Z",
                        "lastTransitionTime": "2024-04-28T03:20:17Z",
                        "reason": "NodeStatusUnknown",
                        "message": "Kubelet stopped posting node status."
                    },
                    {
                        "type": "PIDPressure",
                        "status": "Unknown",
                        "lastHeartbeatTime": "2024-04-28T03:19:05Z",
                        "lastTransitionTime": "2024-04-28T03:20:17Z",
                        "reason": "NodeStatusUnknown",
                        "message": "Kubelet stopped posting node status."
                    },
                    {
                        "type": "Ready",
                        "status": "Unknown", <--------------------------------------------------- node not ready?
                        "lastHeartbeatTime": "2024-04-28T03:19:05Z",
                        "lastTransitionTime": "2024-04-28T03:20:17Z",
                        "reason": "NodeStatusUnknown",
                        "message": "Kubelet stopped posting node status."
                    }
                ],
```

ormergi commented:
Note that only 1/9 of the failing job (of the latest branch) fail for different reason [job](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784271124616450048) 
That is the seems to be the disk0 issue we had while back.

ormergi commented on pkg/virt-operator/resource/apply/apps.go:
```diff
@@ -276,7 +276,7 @@ func (r *Reconciler) syncDaemonSet(daemonSet *appsv1.DaemonSet) (bool, error) {
 	injectOperatorMetadata(kv, &daemonSet.ObjectMeta, imageTag, imageRegistry, id, true)
 	injectOperatorMetadata(kv, &daemonSet.Spec.Template.ObjectMeta, imageTag, imageRegistry, id, false)
-	InjectPlacementMetadata(kv.Spec.Workloads, &daemonSet.Spec.Template.Spec)
+	InjectPlacementMetadata(kv.Spec.Workloads, &daemonSet.Spec.Template.Spec, AnyNode)
```
Similar to the comment about Deployment, here its seems the same for DaemonSets, does it make sense to limit this change to virt-handler only, that should be deployed on worker nodes only?

iholder101 replied:
This is being limited only to daemon sets, and virt-handler is our only daemon set. It is the only exception.

ormergi commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -125,7 +132,7 @@ const (
 
 // Merge all Tolerations, Affinity and NodeSelectos from NodePlacement into pod spec
-func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec) {
+func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev1.PodSpec, nodePlacementOption DefaultInfraComponentsNodePlacement) {
```
nit: It would be nice expressing the intent to pass the node-plac.-opt. to this function in follow up commit in the commit message

iholder101 replied:
sure, I can do that

ormergi commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -136,11 +136,53 @@ func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev
+		case AnyNode:
+			componentConfig = &v1.ComponentConfig{NodePlacement: &v1.NodePlacement{}}
+
+		case RequireControlPlanePreferNonWorker:
```
Please consider exporting the building of `ComponentConfig` and / or its complex-objects to function (e.g: newPlacementRuleAnyNode..) its not easy to follow already, lets not add more to it :slightly_smiling_face:

ormergi commented on pkg/virt-operator/resource/apply/reconcile.go:
```diff
@@ -136,11 +136,53 @@ func InjectPlacementMetadata(componentConfig *v1.ComponentConfig, podSpec *corev
+							},
+							PreferredDuringSchedulingIgnoredDuringExecution: []corev1.PreferredSchedulingTerm{
+								{
+									Weight: 100,
```
Just to understand, is there a chance for this rule to conflict with another with similar weight?

iholder101 replied:
There is no another rule :)
This is only being set by default if the user did not specify any node placements.
I think that in such cases the weight is somewhat redundant.

RamLavi commented:
Hey @iholder101
Thank you for this important change!
I'm curious whether this change will be problematic for SNOs or concrete (not sure this is the right naming) clusters. Did you try?

iholder101 commented:
@ormergi thank you very much for this analysis!
Seems like node02 is labelled with `node.kubernetes.io/unreachable`.

Do you have any idea why this is happening only for the network lanes?

> That is the seems to be the disk0 issue we had while back

Can you elaborate? Which disk0 issue?

iholder101 commented:
> Hey @iholder101 Thank you for this important change! I'm curious whether this change will be problematic for SNOs or concrete (not sure this is the right naming) clusters. Did you try?

Bear in mind that I'm only changing defaults here. SNO can change the node placement and configure it however they please, just like any other environment.

I do need to check if `make cluster-up && make cluster-sync` works with a single node though. I will try that!

iholder101 commented:
> > Hey @iholder101 Thank you for this important change! I'm curious whether this change will be problematic for SNOs or concrete (not sure this is the right naming) clusters. Did you try?
> 
> Bear in mind that I'm only changing defaults here. SNO can change the node placement and configure it however they please, just like any other environment.
> 
> I do need to check if `make cluster-up && make cluster-sync` works with a single node though. I will try that!

Yes it's working :+1: @RamLavi

iholder101 commented:
After speaking with @ormergi offline, he explained that in this [job](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784392104701595648) we can see in node02 [dmesg log](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1784392104701595648/artifacts/k8s-reporter/nodes/1_dmesg_node02.log) its out of memory and kill processes:
```
kern  :info  : [Sun Apr 28 03:40:29 2024] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=system.slice,mems_allowed=0,global_oom,task_memcg=/kubepods/burstable/podf62c7d9d-1b22-47d1-ad9e-a2bf40fc9ed4/crio-2e78c1820d3ef1d5e29e7345b79a61e34934351342af1a3251c8464f950e0d90,task=pilot-discovery,pid=5426,uid=1337
kern  :err   : [Sun Apr 28 03:40:29 2024] Out of memory: Killed process 5426 (pilot-discovery) total-vm:4997792kB, anon-rss:58664kB, file-rss:0kB, shmem-rss:0kB, UID:1337 pgtables:392kB oom_score_adj:989
kern  :info  : [Sun Apr 28 03:40:29 2024] Tasks in /kubepods/burstable/podf62c7d9d-1b22-47d1-ad9e-a2bf40fc9ed4/crio-2e78c1820d3ef1d5e29e7345b79a61e34934351342af1a3251c8464f950e0d90 are going to be killed due to memory.oom.group set
kern  :err   : [Sun Apr 28 03:40:29 2024] Out of memory: Killed process 5426 (pilot-discovery) total-vm:4997792kB, anon-rss:58664kB, file-rss:0kB, shmem-rss:0kB, UID:1337 pgtables:392kB oom_score_adj:989
```

This theoretically makes sense, as now node01 is not tainted anymore (after this PR https://github.com/kubevirt/kubevirtci/pull/1173), so more workloads could be scheduled to it, while all VM workloads must run on node02. This might lead us to rethink the resources we allocate to every node.

However, both @ormergi and I ran the tests offline and they pass consistently.
Before digging any further, I've rebased the PR and now would try to rerun the network lanes.

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network
/test pull-kubevirt-e2e-k8s-1.28-sig-network
/test pull-kubevirt-e2e-k8s-1.27-sig-network

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network
/test pull-kubevirt-e2e-k8s-1.28-sig-network

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network
/test pull-kubevirt-e2e-k8s-1.28-sig-network

iholder101 commented:
Bumping the VM's memory didn't help. And I think I understand why.
Just before the node starts oom-killing, the [logs](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1787150521077862400/artifacts/k8s-reporter/nodes/1_dmesg_node02.log) say: `2359162 pages RAM`.

We can see that `2359162 pages` == `2359162 * 4Ki` == `((2359162 * 4) / (1024^2)) Gi` =~ `9Gi`
That's actually consistent with:
https://github.com/kubevirt/kubevirt/blob/3763f9f2a834e964d3ee84ddb669932347b6f9bb/automation/test.sh#L126

Therefore my conclusion is that the oom-kill is indeed due to a node-level pressure, so the last commit which bumps the VM's memory is unnecessary.

As I've explained [above](https://github.com/kubevirt/kubevirt/pull/11659#issuecomment-2094761205):
> This theoretically makes sense, as now node01 is not tainted anymore (after this PR https://github.com/kubevirt/kubevirtci/pull/1173), so more workloads could be scheduled to it, while all VM workloads must run on node02. This might lead us to rethink the resources we allocate to every node.

@dhiller @brianmcarey can you please share your thoughts here? I can think of a few options:
* Bump the node's resources for network lanes.
  * Better in terms of testing, since it would better reflect a "real life scenario" in which VMs can only run on worker nodes.
* Try to load-balance the cluster using node affinity / selectors / etc (e.g. since all VM workloads run on node02, run other workloads on node01).
	* Better CI resource utilization - since node01 is a control-plane node in every kubevirtCI cluster, it means that we would almost always have 1 CP node which is not fully utilized.
* Change node placement on tests so VMs could run on any node.
	* Best in terms of resource utilization, but the most far from "real-life" behavior.
* Increase the worker node's memory, decrease the CP node's memory.
	* A middle ground: this way we don't have to change node placement or anything like that, and wouldn't have to waste more CI resources.

As you can see there's a trade-off between the test quality, or how close it is to "real life" scenarios, and resource utilization on CI. WDYT?

FYI @EdDev @ormergi

ormergi commented:
On sig-net jobs we deploy istio and CNAO that deploy bunch of other components.
Some CNAO components may no be necessary for kubevirt/kubevirt tests, I think we should revisit then and check which ones are actually needed.
For example OVS CNI is deployed but not tested, the nodes doesnt even have OVS back-end installed.
I posted a PR to set CNAO not deploy OVS-CNI by default https://github.com/kubevirt/kubevirtci/pull/1184

@EdDev @AlonaKaplan can you tell which components are not tested in kubevirt CI? (e.g.: kubeScondaryDNS, dynamicNetworkCotnroller)

iholder101 commented:
Also bear in mind that this PR solves a crucial security issue.
We should prefer fixing the security issue and leave fine-tuning and optimizations to follow-ups as much as reasonable IMO.

iholder101 commented:
> Bumping the VM's memory didn't help. And I think I understand why. Just before the node starts oom-killing, the [logs](https://storage.googleapis.com/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/11659/pull-kubevirt-e2e-k8s-1.29-sig-network/1787150521077862400/artifacts/k8s-reporter/nodes/1_dmesg_node02.log) say: `2359162 pages RAM`.
> 
> We can see that `2359162 pages` == `2359162 * 4Ki` == `((2359162 * 4) / (1024^2)) Gi` =~ `9Gi` That's actually consistent with:
> 
> https://github.com/kubevirt/kubevirt/blob/3763f9f2a834e964d3ee84ddb669932347b6f9bb/automation/test.sh#L126
> 
> Therefore my conclusion is that the oom-kill is indeed due to a node-level pressure, so the last commit which bumps the VM's memory is unnecessary.
> 
> As I've explained [above](https://github.com/kubevirt/kubevirt/pull/11659#issuecomment-2094761205):
> 
> > This theoretically makes sense, as now node01 is not tainted anymore (after this PR [kubevirt/kubevirtci#1173](https://github.com/kubevirt/kubevirtci/pull/1173)), so more workloads could be scheduled to it, while all infra workloads must run on node01. This might lead us to rethink the resources we allocate to every node.
> 
> @dhiller @brianmcarey can you please share your thoughts here? I can think of a few options:
> 
> * Bump the node's resources for network lanes.
>   
>   * Better in terms of testing, since it would better reflect a "real life scenario" in which VMs can only run on worker nodes.
> * Try to load-balance the cluster using node affinity / selectors / etc (e.g. since all VM workloads run on node02, run other workloads on node01).
>   
>   * Better CI resource utilization - since node01 is a control-plane node in every kubevirtCI cluster, it means that we would almost always have 1 CP node which is not fully utilized.
> * Change node placement on tests so VMs could run on any node.
>   
>   * Best in terms of resource utilization, but the most far from "real-life" behavior.
> * Increase the worker node's memory, decrease the CP node's memory.
>   
>   * A middle ground: this way we don't have to change node placement or anything like that, and wouldn't have to waste more CI resources.
> 
> As you can see there's a trade-off between the test quality, or how close it is to "real life" scenarios, and resource utilization on CI. WDYT?
> 
> FYI @EdDev @ormergi

After rethinking the issue I don't understand why node02 suffers from OOM kills and not node01, since it now can both run VMs and run infra workloads.

In addition, all VMs are still being scheduled always to node02 even when node01 has more free space.

I'll investigate it a bit further. Sorry for the noise.

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

EdDev commented:
This one failed migration.

I see in the events:
```json
            "reason": "FailedScheduling",
            "message": "0/2 nodes are available: 2 Insufficient memory. preemption: 0/2 nodes are available: 2 No preemption victims found for incoming pod.",
            "source": {
                "component": "default-scheduler"
            },
```

So even without the passt, it would not have worked.
I suspect the problem can be solved by adding a 3rd node.
But I am unsure about the implications on the CI load.

We can try increasing the nodes just to see if it works.
And if we do that, I would suggest not to run workloads on the control node.

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

iholder101 commented:
It passed!! :tada: 

> So even without the passt, it would not have worked. I suspect the problem can be solved by adding a 3rd node. But I am unsure about the implications on the CI load.
> 
> We can try increasing the nodes just to see if it works. And if we do that, I would suggest not to run workloads on the control node.

@dhiller WDYT about adding another node for the network lanes as @EdDev suggested?

Barakmor1 commented:
> It passed!! 🎉
> 
> > So even without the passt, it would not have worked. I suspect the problem can be solved by adding a 3rd node. But I am unsure about the implications on the CI load.
> > We can try increasing the nodes just to see if it works. And if we do that, I would suggest not to run workloads on the control node.
> 
> @dhiller WDYT about adding another node for the network lanes as @EdDev suggested?

Would increasing the memory of the worker node also work?

iholder101 commented:
> > It passed!! 🎉
> > > So even without the passt, it would not have worked. I suspect the problem can be solved by adding a 3rd node. But I am unsure about the implications on the CI load.
> > > We can try increasing the nodes just to see if it works. And if we do that, I would suggest not to run workloads on the control node.
> > 
> > 
> > @dhiller WDYT about adding another node for the network lanes as @EdDev suggested?
> 
> Do we need to do this for all lanes as done in this PR, or only the network ones?

Only the network lanes.
Never mind my commit, it's just a quick-and-dirty way to check if the network lanes pass.

iholder101 commented:
@dhiller Also, would the right place to do so be `test.sh` (like done in this PR, but only for network lanes) or in project infra, e.g. [here](https://github.com/kubevirt/project-infra/blob/e19086df9af8b5259864a9aefa022f7d3e59767f/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml#L1602-L1614):
```go
    name: pull-kubevirt-e2e-k8s-1.29-sig-network
    skip_branches:
    - release-\d+\.\d+
    spec:
      containers:
      - command:
        - /usr/local/bin/runner.sh
        - /bin/sh
        - -c
        - automation/test.sh
        env:
        - name: TARGET
          value: k8s-1.29-sig-network
```

iholder101 commented:
/test pull-kubevirt-e2e-k8s-1.29-sig-network

iholder101 commented:
/test all

iholder101 commented:
/retest

/test pull-kubevirt-e2e-k8s-1.28-sig-compute pull-kubevirt-e2e-k8s-1.28-sig-network pull-kubevirt-e2e-k8s-1.28-sig-operator pull-kubevirt-e2e-k8s-1.28-sig-storage pull-kubevirt-e2e-k8s-1.29-sig-compute pull-kubevirt-e2e-k8s-1.29-sig-operator pull-kubevirt-e2e-k8s-1.29-sig-storage pull-kubevirt-e2e-k8s-1.30-ipv6-sig-network pull-kubevirt-e2e-kind-sriov pull-kubevirt-e2e-windows2016

iholder101 commented:
All lanes are passing now :)
/unhold

@fossedihelm can PTAL?

iholder101 commented:
/uncc @0xFelix @AlonaKaplan @lyarwood

fossedihelm reviewed (approved):
Thank you!
